### PR TITLE
Archify EthStateCache

### PR DESCRIPTION
**Pull request by i-m-aditya**
Ref #5720

### CODE REVIEW

File: crates/rpc/rpc/src/eth/cache/mod.rs [hunk 1]
```diff
@@ -413,8 +416,9 @@ where
                         }
                         CacheAction::GetReceipts { block_hash, response_tx } => {
                             // check if block is cached
-                            if let Some(receipts) = this.receipts_cache.get(&block_hash).cloned() {
-                                let _ = response_tx.send(Ok(Some(receipts)));
+                            if let Some(receipts) = this.receipts_cache.get(&block_hash) {
+                                let receipts = Arc::new(receipts.clone());
```
Comment (i-m-aditya) [inline crates/rpc/rpc/src/eth/cache/mod.rs:420]: @mattsse Is this the right way or should I archify it `this.receipts_cache.get(&block_hash)` function call itself.

File: crates/rpc/rpc/src/eth/api/block.rs [hunk 2]
```diff
@@ -87,7 +87,7 @@ where
             let receipts = block
                 .body
                 .into_iter()
-                .zip(receipts.clone())
+                .zip(receipts.to_vec())
```
Comment (i-m-aditya) [inline crates/rpc/rpc/src/eth/api/block.rs:90]: I was sceptic using this, any other way to not use extra memory and a workaround only using arc.

**Review by rkrasiuk — DISMISSED** [review]
couple of comments:
1. We should not change provider interfaces for this purpose
2. The values should be wrapped in `Arc` inside lru cache itself and the `Arc` should only be created upon insertion
3. Having values wrapped in an `Arc` would require creating a custom limiter since `ByLength` is not implemented for `Arc<Vec<T>>`
4. All of the other `to_vec` and `clone` calls should become unnecessary after completion of all steps above

**Comment by i-m-aditya** [comment]
> couple of comments:
> 
> 1. We should not change provider interfaces for this purpose
> 2. The values should be wrapped in `Arc` inside lru cache itself and the `Arc` should only be created upon insertion
> 3. Having values wrapped in an `Arc` would require creating a custom limiter since `ByLength` is not implemented for `Arc<Vec<T>>`
> 4. All of the other `to_vec` and `clone` calls should become unnecessary after completion of all steps above

@rkrasiuk Thanks for the comments. However, I have one question
`let mut block_and_receipts = None;

        if block_id.is_pending() {
            block_and_receipts = self.provider().pending_block_and_receipts()?;
        } else if let Some(block_hash) = self.provider().block_hash_for_id(block_id)? {
            block_and_receipts = self.cache().get_block_and_receipts(block_hash).await?;
        }`
       For block_and_receipts to align with the same type, it can be Arc or the value itself. For this to happen, I need to change the provider interface. Am I missing something here?

**Comment by mattsse** [comment]
ah I see,
need to think about more, but we should proceed with simply putting the pending receipts in a new Arc before returning

**Comment by i-m-aditya** [comment]
> ah I see, need to think about more, but we should proceed with simply putting the pending receipts in a new Arc before returning

Yeah, would wrap the receipt in Arc and not change the provider interface.

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**